### PR TITLE
Maria Velez - Octos

### PR DESCRIPTION
**Pull request by velezdev**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did you go about exploring the Edamam API, how did you try querying the API? | Once I had my Edamam ID and Key, I used Postman to query the API using different parameters
Describe your API Wrapper.  How did you decide on the methods you created? | I created two methods: self.one_recipe that will be used to show one recipe view and self.all_recipes to view the search results after a query (what I called "user_input") was entered. I thought it best to split the functionality to these two methods instead of creating it in one.
Describe an edge case or failure case test you wrote for your API Wrapper. | My test called 'responds with error with an invalid search' tested an empty query search.
Explain how VCR aids in testing an API. | The VCR wraps the code where an API call would be made so that a request to the API is not made every single time we run a test.
What is the Heroku URL of your deployed application? | I did not deploy to Heroku as of this PR (per Kari's instructions)
Provide a link to the Trello board you used | https://trello.com/b/XqksJIvr/api-muncher-maria-velez-ada-c9-octos

### CODE REVIEW

File: app/controllers/recipes_controller.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
+class RecipesController < ApplicationController
+
+  def homepage; end
+
+  def index
+    recipes = EdamamApiWrapper.all_recipes(params[:user_input])
+
+    @recipes = Kaminari.paginate_array(recipes).page(params[:page]).per(12)
+
+    if @recipes == []
```
Comment (kariabancroft) [inline app/controllers/recipes_controller.rb:10]: I like the way you're providing additional information when there are no recipes found. I think that the `all_recipes` method can still throw an error that you wouldn't handle and/or share with your user.

File: lib/edamam_api_wrapper.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  APP_KEY = ENV['EDAMAM_KEY']
+
+  def self.one_recipe(uri)
+    # this is linked to show page
+    url = "#{BASE_URL}r=http://www.edamam.com/ontologies/edamam.owl#recipe_#{uri}&app_id=#{APP_ID}&app_key=#{APP_KEY}"
+
+    encoded_url = URI.encode(url)
+    response = HTTParty.get(encoded_url)
+
+    if response[0].nil?
```
Comment (kariabancroft) [inline lib/edamam_api_wrapper.rb:17]: I like that you're checking this value prior to attempting to send this data through to create a Recipe instance.

File: app/views/recipes/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,29 @@
+
+<h1 class="recipe-label-show"><%= @recipe.label %></h1>
```
Comment (kariabancroft) [inline app/views/recipes/show.html.erb:2]: You return `nil` from the API wrapper's method if the data isn't found, but if `@recipe` equals `nil` within this view, you'll have issues trying to retrieve these properties. I'd recommend adding some additional logic to the controller to handle this value instead.

File: lib/edamam_api_wrapper.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
+require 'httparty'
+
+class EdamamApiWrapper
+  class EdamamError < StandardError; end
+
+  BASE_URL = "https://api.edamam.com/search?"
+  APP_ID = ENV['EDAMAM_ID']
+  APP_KEY = ENV['EDAMAM_KEY']
+
+  def self.one_recipe(uri)
```
Comment (kariabancroft) [inline lib/edamam_api_wrapper.rb:10]: Is there any checking that should be done on the `uri` before it goes into a request?

File: lib/edamam_api_wrapper.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    response = HTTParty.get(encoded_url)
+
+    if response[0].nil?
+      return nil
+    else
+      return Recipe.from_api(response[0])
+    end
+  end
+
+  def self.all_recipes(user_input)
```
Comment (kariabancroft) [inline lib/edamam_api_wrapper.rb:24]: Is there any checking that should be done on this search term before passing it through to the request?

File: test/controllers/recipes_controller_test.rb [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    it "can get the index path" do
+      VCR.use_cassette("recipes") do
+        get recipes_path, params: {user_input: "avocado"}
+        must_respond_with :success
+      end
+    end
+
+    it 'fails with nonsensical input' do
+      VCR.use_cassette("recipes") do
+        proc {get recipes_path, params: {user_input: "thesetestsarenotpassing"} }. must_raise
```
Comment (kariabancroft) [inline test/controllers/recipes_controller_test.rb:19]: Seems like you're missing the parameter to be passed to `must_raise` here

File: test/controllers/recipes_controller_test.rb [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+      end
+    end
+
+    it 'succeeds with valid data' do
```
Comment (kariabancroft) [inline test/controllers/recipes_controller_test.rb:23]: I think this test is the same as the test above "can get the index path"

File: test/controllers/recipes_controller_test.rb [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+        get recipe_path(uri)
+        must_respond_with :success
+      end
+    end
+
+    it "fails with an invalid search" do
+      VCR.use_cassette("recipes") do
+        proc {
+          get recipe_path(thiswillnotwork) }.must_raise
```
Comment (kariabancroft) [inline test/controllers/recipes_controller_test.rb:44]: Same comment re: the `must_raise` param

File: test/lib/edamam_api_wrapper_test.rb [hunk 9]
```diff
@@ -0,0 +1,30 @@
+require 'test_helper'
+
+describe EdamamApiWrapper do
+  describe 'one_recipe' do
+    it 'returns one recipe' do
+      VCR.use_cassette('recipe') do
+        recipe = EdamamApiWrapper.one_recipe("b1a783e4ad64f147e4bf3344464602c6")
```
Comment (kariabancroft) [inline test/lib/edamam_api_wrapper_test.rb:7]: What about a test with a bad uid?

File: lib/edamam_api_wrapper.rb [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    url = "#{BASE_URL}q=#{user_input}&app_id=#{APP_ID}&app_key=#{APP_KEY}&from=0&to=999"
+
+    encoded_url = URI.encode(url)
+    response = HTTParty.get(encoded_url)
+    # in case of a bad ID or Key, for example
+    unless response["hits"]
+      raise
```
Comment (kariabancroft) [inline lib/edamam_api_wrapper.rb:31]: Oh no! Not a `raise` in the submitted code!

**Comment by kariabancroft** [comment]
# API Muncher
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Core Requirements** | 
Git hygiene | Good
Comprehension questions | Yes
**General** | 
Rails fundamentals (RESTful routing, use of named paths) | Yes
Semantic HTML | Yes - you did a nice job organizing your HTML into appropriate sections/groups
Errors are reported to the user | Some - some errors with the list of recipes, and some handling for bad data from the API, but missing the thread that takes that all the way through
API Wrapper to handle the API requests | Yes
Controller testing | Yes - though it looks like the tests for the error cases won't work
Lib testing | Yes - again missing error cases
Search Functionality | Yes - though it doesn't work right now
List Functionality | Yes - though can't verify it is working
Show individual item functionality (link to original recipe opens in new tab) | Yes
**Styling** | 
Responsive layout | Yes
List View shows 10 items at a time/pagination | Yes - using Kanimari gem
The app is styled to create an attractive user interface | Yes
**API Features** | 
The App attributes Edaman | Yes
The VCR casettes do not contain the API key | Correct - nice job!
**External Resources** | 
Link to Trello Board | Yes - nice organization on this!
Link to deployed app on Heroku | N/A
**Overall** | Overall reviewing your code showed lots of good examples of integrating the API through the wrapper, lib class and controller. It should always be assumed that we'll run your code when you submit, and it seems like you've left some extraneous code in here that prevents things from working. Overall, your positive test cases look good, but you've left room for improvement on the error or edge cases.
